File: stacks-queues/queue.js
class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

class Queue {
  constructor() {
    this.front = null;
    this.rear = null;
    this.length = 0;
  }

  enqueue(value) {
    let newNode = new Node(value);

    if (this.isEmpty()) {
      this.front = this.rear = newNode;
    } else {
      this.rear.next = newNode;
      this.rear = newNode;
    }
    this.length++;
  }

  dequeue() {
    if (this.isEmpty()) {
      return null;
    }
    let value = this.front.value;
    this.front = this.front.next;
    this.length--;

    if (this.isEmpty()) {
      this.rear = null;
    }

    return value;
  }

  // Check if the queue is empty
  isEmpty() {
    return this.length === 0;
  }

  // Return the value of the front item
  peek() {
    return this.isEmpty() ? null : this.front.value;
  }

  // Return the size of the queue
  size() {
    return this.length;
  }
}

File: stacks-queues/queue.py
class Node:
    def __init__(self, value, next = None):
        self.value = value
        self.next = next

class Queue:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def enqueue(self, value):
        new_tail = Node(value)

        if self.is_empty():
            self.head = new_tail
        else:
            self.tail.next = new_tail

        self.tail = new_tail
        self.size += 1

        return self

    def dequeue(self):
        if self.is_empty():
            return None

        value = self.head.value
        self.head = self.head.next
        self.size -= 1

        if self.is_empty():
            self.tail = None

        return value

    def is_empty(self):
        return self.size == 0

    def peek(self):
        return None if self.is_empty() else self.head.value
